src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/dynamicProgramming/DynamicProgramming.java

```diff
@@ -0,0 +1,74 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.dynamicProgramming;
+
+import se.liu.ida.hefquin.engine.federation.FederationAccessException;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpJoin;
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.*;
+import se.liu.ida.hefquin.engine.queryproc.*;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CardinalityEstimation;
+
+import java.util.*;
+
+public class DynamicProgramming extends CardinalityEstimation {
+    protected final QueryProcContext ctxt;
+    protected List<PhysicalPlan> lpList;
+
+    public DynamicProgramming(final QueryProcContext ctxt, final List<PhysicalPlan> lpList) {
+        assert ctxt != null;
+        this.ctxt = ctxt;
+        this.lpList = lpList;
+    }
+
+    protected PhysicalPlan optimizePhysicalPlanForMultiwayJoin() throws FederationAccessException {
+        PhysicalPlan currentPlan = chooseFirstSubquery();
+
+        while (lpList.size() > 0){
+            currentPlan = cardinalityTwoSubQueries(currentPlan);
+        }
+
+        return currentPlan;
+    }
+
+    public PhysicalPlan chooseFirstSubquery() throws FederationAccessException {
```

Comment:
QueryOptimizationException

Reply:
Also, this method should be protected rather than public. You don't want it to be called from the outside.